Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx

```diff
@@ -0,0 +1,382 @@
+  bool isEditing();
+  void setEditing(bool editing);
+};
+
+// --------------------------------------------------------------------------
+qMRMLTextWidgetPrivate::qMRMLTextWidgetPrivate(qMRMLTextWidget& object)
+  : q_ptr(&object)
+  , CurrentTextNode(nullptr)
+  , TextEditModified(false)
+  , TextNodeModified(false)
+  , Editing(false)
+{
+}
+
+//-----------------------------------------------------------------------------
+qMRMLTextWidgetPrivate::~qMRMLTextWidgetPrivate() = default;
+
+// --------------------------------------------------------------------------
+void qMRMLTextWidgetPrivate::setupUi(qMRMLTextWidget* widget)
+{
+  this->Ui_qMRMLTextWidget::setupUi(widget);
+}
+
+// --------------------------------------------------------------------------
+bool qMRMLTextWidgetPrivate::isEditing()
+{
+  return this->Editing;
+}
+
+// --------------------------------------------------------------------------
+void qMRMLTextWidgetPrivate::setEditing(bool editing)
+{
+  Q_Q(qMRMLTextWidget);
+  if (this->Editing == editing)
+    {
+    return;
+    }
+
+  this->Editing = editing;
+  q->editingChanged(this->Editing);
+}
+
+//-----------------------------------------------------------------------------
+// qMRMLTextWidget methods
+
+//-----------------------------------------------------------------------------
+qMRMLTextWidget::qMRMLTextWidget(QWidget* parentWidget)
+  : Superclass( parentWidget )
+  , d_ptr(new qMRMLTextWidgetPrivate(*this))
+  , AutoUpdate(false)
+  , ReadOnly(false)
+{
+  this->setup();
+}
+
+//-----------------------------------------------------------------------------
+qMRMLTextWidget::~qMRMLTextWidget()
+{
+  this->setMRMLTextNode(nullptr);
+}
+
+//-----------------------------------------------------------------------------
+void qMRMLTextWidget::setup()
+{
+  Q_D(qMRMLTextWidget);
+
+  d->setupUi(this);
+
+  connect(d->TextEdit, SIGNAL(textChanged()), this, SLOT(onTextEditChanged()));
+  connect(d->EditButton, SIGNAL(clicked()), this, SLOT(onEditButtonClicked()));
+  connect(d->CancelButton, SIGNAL(clicked()), this, SLOT(onCancelButtonClicked()));
+  connect(d->SaveButton, SIGNAL(clicked()), this, SLOT(onSaveButtonClicked()));
+  connect(this, SIGNAL(updateRequested()), this, SLOT(updateWidget()));
+
+  this->updateRequested();
+}
+
+//------------------------------------------------------------------------------
+void qMRMLTextWidget::setMRMLNode(vtkMRMLNode* node)
+{
+  this->setMRMLTextNode(vtkMRMLTextNode::SafeDownCast(node));
+}
+
+//------------------------------------------------------------------------------
+void qMRMLTextWidget::setMRMLTextNode(vtkMRMLTextNode* node)
+{
+  Q_D(qMRMLTextWidget);
+  if (node == d->CurrentTextNode)
+    {
+    // not changed
+    return;
+    }
+
+  // Reconnect the appropriate nodes
+  this->qvtkReconnect(d->CurrentTextNode, node, vtkMRMLTextNode::TextModifiedEvent, this, SLOT(onTextNodeModified()));
+  d->CurrentTextNode = node;
+  d->Editing = false;
+  this->updateRequested();
```

Comment:
Direction of update is not clear. Use updateWidgetFromMRML (updateWidgetFromMRMLRequested).